z3c/dependencychecker/db.py

```diff
@@ -43,3 +48,186 @@ def _filter_duplicates(self, imports):
+            imports_list=should_be_test_requirements,
+        )
+        return unique
+
+    def get_unneeded_test_requirements(self):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return []
+
+        filters = (
+            self._filter_known_packages,
+            self._filter_python_standard_library,
+            self._filter_used_imports,
+        )
+        unneeded = self._process_pipeline(test_requirements, filters)
+        unique_imports = self._get_unique_imports(imports_list=unneeded)
+        return unique_imports
+
+    def _process_pipeline(self, objects, filters):
+        final_list = []
+        for single_object in objects:
+            for filter_func in filters:
+                if not filter_func(single_object):
+                    break
+            else:
+                final_list.append(single_object)
+
+        return final_list
+
+    def _get_unique_imports(self, imports_list=None):
+        if imports_list is None:
+            imports_list = self.imports_used
+        unique_dotted_names = list(set(imports_list))
+        sorted_unique_dotted_names = sorted(unique_dotted_names)
+        return sorted_unique_dotted_names
+
+    def _filter_used_imports(self, dotted_name):
+        return self._filter_requirements_lists(self.imports_used, dotted_name)
+
+    def _filter_known_packages(self, dotted_name):
+        to_ignore = (
+            DottedName('setuptools'),
+            DottedName('pkg_resources'),
+            DottedName('distribute'),
+        )
+        return self._filter_requirements_lists(to_ignore, dotted_name)
+
+    @staticmethod
+    def _filter_testing_imports(dotted_name):
+        return not dotted_name.is_test
+
+    @staticmethod
+    def _filter_only_testing_imports(dotted_name):
+        return dotted_name.is_test
+
+    def _filter_own_package(self, dotted_name):
+        own_package = self._build_own_package_import()
+        return dotted_name not in own_package
+
+    def _build_own_package_import(self):
+        return DottedName(self.package_name)
+
+    def _filter_requirements(self, dotted_name):
+        return self._filter_requirements_lists(
+            self._requirements,
+            dotted_name,
+        )
+
+    def _filter_test_requirements(self, dotted_name):
+        test_requirements = self._get_test_extra()
+        if not test_requirements:
+            return True
+
+        return self._filter_requirements_lists(
+            test_requirements,
+            dotted_name,
+        )
+
+    def _filter_python_standard_library(self, dotted_names):
+        std_library = self._build_std_library()
+        return self._filter_requirements_lists(std_library, dotted_names)
+
+    @staticmethod
+    def _filter_requirements_lists(requirements, candidate):
+        return not any((r for r in requirements if candidate in r))
+
+    @staticmethod
+    def _build_std_library():
+        python_version = sys.version_info
```

Comment:
This makes the result of the dependencychecker output dependent on the python version, but that is not something that can be helped.
Perhaps we should add a note in the README later on.
OTOH, it ought to be clear enough for people using this tool, right?

Reply:
Now that you mention it... good question!

Given that z3c.dependencychecker can be ``pip install``ed, one would expect that it uses the same python version as the rest of the code installed on that virtual environment...

**But** what about code that tries to be compatible with Python 2 and Python 3? On Python 3 there's quite a few new/removed modules on the standard library, so when running on py2 you will get warnings for the new ones, and the other way around...

Should we use #41 as a way to signal which versions of python standard library to check against, and fallback to the current python version?

Reply:
I think we don't have to do anything fancy, after all. Dependencychecker is aimed at a quite advanced audience, so they'll know it when they see an anomaly from a python2/3 difference. Perhaps a line in the readme, that will be enough.

Reply:
An alternative would be to just add the 2.7 and 3.6 dependencies in one big list. That way we'll even handle "except ImportError" conditional imports to handle 2/3 issues. Would that be an idea?